### ISSUE TITLE
[Windows Hi-DPI] Progress icon is too big and clipped

### ISSUE THREAD

**Issue by Phillipus** [eclipse-platform/eclipse.platform.swt#706]
Eclipse 4.28
Windows 11
Hi-DPI screen 200% scaling

When running an operation that uses Progress, the icon in the status bar (progress bar?) is too big and so we only see a quarter of it:

![progress](https://github.com/eclipse-platform/eclipse.platform.swt/assets/600504/d6f486b9-2c7a-4f57-b86f-7d3ce846a35f)

**Comment by tmssngr** [comment]
@Phillipus Have you already tried running Eclipse with **swt.autoScale** set to **false**? We do that for SmartGit and it solves a couple of problems.

**Comment by merks** [comment]
There's also this setting on the java executable (javaw.exe) that has an impact:

![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/208716/16d18a92-75ec-47f1-bfd9-cc7d25c00f6a)

I make sure to change it to system rather than application so that Windows does the scaling and not SWT which does a horrible job.

**Comment by Phillipus** [comment]
I think this is a case that was missed when creating images at double-scale on Windows. I'll have to dig deeper but I think it used to work.

**Comment by Phillipus** [comment]
The icon is the correct size in Eclipse SDK 4.13 (Sep 2019), not shown at all in the progress bar in 4.14 (Dec 2019), 4,15 (Mar 2020) and then shows the wrong size in 4.16 (June 2020).

**Comment by Phillipus** [comment]
> but I think it used to work.

It does and it doesn't appear at the correct size in previous versions of Eclipse. It depends whether that "Tip of the day" icon is showing in the bottom status bar or not. Sometimes the progress icon doesn't appear at all.

**Comment by Phillipus** [comment]
This is connected to another issue I see in Eclipse at 200% scaling.

Sometimes the icons used for files (Program icons) in the Package Explorer have a black background and sometimes they don't. If they don't have a black background the progress icon is rendered correctly.

See https://github.com/eclipse-platform/eclipse.platform.swt/issues/185

**Comment by HeikoKlare** [comment]
I just noticed that I sometimes see this for other elements in the status bar as well, e.g. 
<img width="260" height="58" alt="Image" src="https://github.com/user-attachments/assets/6c35b2d6-fef3-404d-ada0-7aa23c8b66e6" />

So maybe this is a general layouting issue of the [StatusLine](https://github.com/eclipse-platform/eclipse.platform.ui/blob/a3fa06583905e66550c0b4e8a3961320cfb4ca43/bundles/org.eclipse.jface/src/org/eclipse/jface/action/StatusLine.java#L65) in HiDPI setups. The shown progress icon is defined in `ProgressAnimationItem`.

**Comment by Phillipus** [comment]
In the case of the Progress icon it's 2x too big on HiDPI (and so clipped so we only see 1/4 of it), so I always thought it was either using the wrong icon size or rendering the wrong size.

**Comment by HeikoKlare** [comment]
Yes, the icon in the screenshots appears to have twice the width/height it should have. But the icons themselves all look correctly scaled (16x16 in their 100%/svg version and 32x32 in their 200% version). The image is only [set at on a simple ToolItem](https://github.com/eclipse-platform/eclipse.platform.ui/blob/a3fa06583905e66550c0b4e8a3961320cfb4ca43/bundles/org.eclipse.ui.workbench/eclipseui/org/eclipse/ui/internal/progress/ProgressAnimationItem.java#L263-L267), which is why this does not seem to be so specific to the progress case.

**Comment by arunjose696** [comment]
The issue happens because on Windows, when a toolbar is created without a size, Windows gives each button a default size of 24×22. Our icons are 16×16, so at 100% scaling everything looks fine as it wont be cut off.

On HiDPI screens, the images that get set later are bigger (e.g., 32×32). The toolbar and its buttons resize to fit the bigger image, but the parent  (grid layout) doesn't relayout, so the space stays too small and the icon gets clipped.

This issue of parent not relayouting seems to come from the deferred layout call introduced in this commit:
https://github.com/eclipse-platform/eclipse.platform.ui/commit/84b4e32ecec1772a97e8adabaa257b86c46b2eb3 (which was added for performance reasons: https://bugs.eclipse.org/bugs/show_bug.cgi?id=528030).
Because of this deferred layout, the parent layout doesn't update after the images are set, until a setLayoutDeferred(false) happens  which can occur on something like a DPI change.

A fix is to layout the parent of the toolbar. COnsidering perfromance we could just trigger a relayout only when the image size actually changes, e.g.: when the new image is larger or smaller than the old one, or when changing from null to image.
This matches what SWT already does for  toolbars ie when a ToolItem's image size changes (or goes from null), the toolbar gets relayouted. So doing the same also for the tolbars parent makes sense to prevent it from being cut off.  I have opened a PR with this fix

**Comment by Phillipus** [comment]
> [akoch-yatta](https://github.com/akoch-yatta) closed this as [completed](https://github.com/eclipse-platform/eclipse.platform.swt/issues?q=is%3Aissue%20state%3Aclosed%20archived%3Afalse%20reason%3Acompleted) in [eclipse.platform.ui#3567](https://github.com/eclipse-platform/eclipse.platform.ui/pull/3567)[26 minutes ago](https://github.com/eclipse-platform/eclipse.platform.swt/issues/706#event-21666568640)

@akoch-yatta Thank-you. :-)
